File: src/recovery.ts
/**
 * Unsaved work, and how far it has got.
 *
 * alabs in Chrome may not enable editing without this. A buffer in a browser
 * tab is the least durable place text can be: the tab can be closed, the page
 * reloaded, the browser quit, the machine restarted. So every edit becomes a
 * *draft* — a numbered revision of one file's buffer — and a draft is written
 * twice: into the browser's own storage, which is immediate, and into a file
 * under alabs' state folder, which survives clearing browser data.
 *
 * The rules this module exists to keep:
 *
 *   * A revision is acknowledged only once its storage has actually accepted
 *     it. "Written" is never assumed from "asked to write".
 *   * A write that finishes late acknowledges **its own** revision and no
 *     other. It can never speak for the text that was typed after it.
 *   * Storage that fails says so. The live text stays in hand, and alabs can
 *     hand it to the user, rather than a quiet failure that looks like safety.
 *   * A draft is forgotten only when the file has been saved at that exact
 *     revision, or the user discarded it. Never to make room, never on a
 *     guess.
 *
 * Everything here is plain values and promises, so the ordering rules can be
 * tested without a browser, a server, or a disk.
 */
import type { FileStamp } from "./subject";

/** One revision of one file's unsaved buffer. */
export interface Draft {
  /** Canonical absolute path of the alabs root the file is in. */
  root: string;
  /** The file's path inside that root. */
  relPath: string;
  /** The file's stamp when the buffer last matched the disk; null when the file was already gone. */
  stamp: FileStamp | null;
  contents: string;
  /** Only ever increases for one file. */
  revision: number;
  /** The `alabs-serve` lifetime this was typed against, when there was one. */
  serverId: string | null;
  /** The alabs window that typed it. */
  sessionId: string;
  /** Milliseconds since the epoch. */
  updatedAt: number;
}

/** What one listing found, including records it could not read. */
export interface DraftListing {
  drafts: Draft[];
  /** Names of stored drafts that could not be read. Reported, never dropped. */
  unreadable: string[];
}

/** Somewhere a draft can be kept. Both places alabs keeps them look like this. */
export interface DraftStore {
  /** Store one revision. Rejects rather than storing an older one over a newer. */
  put(draft: Draft): Promise<void>;
  list(root: string): Promise<DraftListing>;
  /** Forget a draft, but only up to `revision`; newer text is kept. */
  drop(root: string, relPath: string, revision: number): Promise<void>;
}

/** How far the newest revision of one draft has got. */
export type Kept =
  /** Typed, and not yet accepted anywhere. */
  | "pending"
  /** Accepted by the browser's own storage. Survives a reload and a closed tab. */
  | "browser"
  /** Also written to alabs' state folder. Survives clearing browser data. */
  | "disk";

/** Where a draft could not be kept, and what was said about it. */
export interface Trouble {
  where: "browser" | "disk";
  reason: string;
}

/** What alabs knows about one file's unsaved work right now. */
export interface DraftState {
  revision: number;
  kept: Kept;
  /** Set while the newest revision could not be kept somewhere. */
  trouble: Trouble | null;
}

export interface RecorderIo {
  /** The alabs root every draft belongs to. */
  root: string;
  /** This `alabs-serve` lifetime, when there is one. */
  serverId: string | null;
  /** This alabs window. */
  sessionId: string;
  /** The browser's own storage: immediate, and lost with the browser's data. */
  browser: DraftStore;
  /** alabs' state folder, through the server: the durable copy. */
  disk: DraftStore;
  now?: () => number;
}

/**
 * Keeps one root's drafts, one revision at a time per file.
 *
 * Writes are single-file: while one revision of a file is being stored, newer
 * text simply replaces what is waiting, so a burst of typing costs two writes
 * rather than one per keystroke, and the newest text is always the one that
 * ends up stored.
 */
export class Recorder {
  private readonly io: RecorderIo;
  /** The newest revision assigned per file. */
  private readonly latest = new Map<string, number>();
  /** The newest revision each store has actually accepted, per file. */
  private readonly acked = new Map<string, { browser: number; disk: number }>();
  /** The newest draft not yet stored, per file. */
  private readonly waiting = new Map<string, Draft>();
  /** The store in progress per file, so settling can wait for it. */
  private readonly storing = new Map<string, Promise<void>>();
  /**
   * The revision each file has been dealt with through: saved, or discarded.
   *
   * Kept for the life of this window rather than cleared with the draft,
   * because it is also the floor the next revision has to clear. Without it, a
   * file saved and then edited again would start at 1, and a store that still
   * held revision 7 — because the drop was refused — would refuse every write
   * from then on.
   */
  private readonly settled = new Map<string, number>();
  /** The newest text per file, kept so it can be handed over if storing fails. */
  private readonly live = new Map<string, Draft>();
  private readonly troubles = new Map<string, Trouble>();
  /** Called whenever a file's `DraftState` changes, so the UI can follow it. */
  changed: (relPath: string) => void = () => {};

  constructor(io: RecorderIo) {
    this.io = io;
  }

  private now(): number {
    return (this.io.now ?? Date.now)();
  }

  private acksFor(relPath: string): { browser: number; disk: number } {
    let acks = this.acked.get(relPath);
    if (!acks) {
      acks = { browser: 0, disk: 0 };
      this.acked.set(relPath, acks);
    }
    return acks;
  }

  /**
   * Start this file's revisions above what was already stored for it.
   *
   * Revisions must only ever increase, and a restored draft was numbered by an
   * earlier alabs window. Without this, the first edit after a restore would
   * be numbered 1 and be refused as older than what is kept.
   */
  seed(relPath: string, revision: number, where: "browser" | "disk"): void {
    this.latest.set(relPath, Math.max(revision, this.latest.get(relPath) ?? 0));
    const acks = this.acksFor(relPath);
    // A listing proves only what that store holds. The other copy may be
    // older, missing, or unreadable; never promote it to an acknowledgement.
    acks[where] = Math.max(acks[where], revision);
  }

  /**
   * Record the buffer of `relPath` as its next revision and start storing it.
   * Text identical to the newest revision is not renumbered: it is already
   * exactly what is being kept.
   *
   * Returns the revision this text is known by, which is what `settle` must
   * later be given.
   */
  note(relPath: string, contents: string, stamp: FileStamp | null): number {
    const current = this.live.get(relPath);
    if (current && current.contents === contents && sameStamp(current.stamp, stamp)) {
      return current.revision;
    }
    const revision = Math.max(this.latest.get(relPath) ?? 0, this.settled.get(relPath) ?? 0) + 1;
    const draft: Draft = {
      root: this.io.root,
      relPath,
      stamp,
      contents,
      revision,
      serverId: this.io.serverId,
      sessionId: this.io.sessionId,
      updatedAt: this.now(),
    };
    this.latest.set(relPath, revision);
    this.live.set(relPath, draft);
    this.waiting.set(relPath, draft);
    this.changed(relPath);
    this.start(relPath);
    return revision;
  }

  /** Begin storing this file's waiting revisions, if nothing is already. */
  private start(relPath: string): void {
    if (this.storing.has(relPath)) return;
    const run = this.store(relPath);
    this.storing.set(relPath, run);
    void run.then(() => {
      this.storing.delete(relPath);
      // Text typed in the moment between the loop ending and this bookkeeping
      // would otherwise sit unwritten until the next keystroke.
      if (this.waiting.has(relPath)) this.start(relPath);
    });
  }

  /** Store whatever is waiting for `relPath`, one revision at a time. */
  private async store(relPath: string): Promise<void> {
    for (;;) {
      const draft = this.waiting.get(relPath);
      if (!draft) return;
      this.waiting.delete(relPath);
      // Dealt with while it waited — saved, or thrown away — so writing it
      // now would put back a draft that is finished with.
      if (draft.revision <= (this.settled.get(relPath) ?? 0)) continue;
      let trouble: Trouble | null = null;
      // The browser first: it is the storage alabs can reach without leaving
      // the page, so it is the one that answers soonest.
      try {
        await this.io.browser.put(draft);
        this.ack(relPath, "browser", draft.revision);
      } catch (err) {
        trouble = { where: "browser", reason: String(err) };
      }
      // The disk regardless: browser storage failing is exactly when the
      // durable copy matters most, and the two do not depend on each other.
      try {
        await this.io.disk.put(draft);
        this.ack(relPath, "disk", draft.revision);
      } catch (err) {
        trouble = { where: "disk", reason: String(err) };
      }
      this.setTrouble(relPath, trouble);
    }
  }

  /**
   * One store accepted one revision. Only ever that revision: a write that
   * finished late must not acknowledge the text typed after it.
   */
  private ack(relPath: string, where: "browser" | "disk", revision: number): void {
    const acks = this.acksFor(relPath);
    if (revision <= acks[where]) return;
    acks[where] = revision;
    this.changed(relPath);
  }

  private setTrouble(relPath: string, trouble: Trouble | null): void {
    const had = this.troubles.get(relPath) ?? null;
    if (trouble === null) {
      if (had === null) return;
      this.troubles.delete(relPath);
    } else {
      if (had && had.where === trouble.where && had.reason === trouble.reason) return;
      this.troubles.set(relPath, trouble);
    }
    this.changed(relPath);
  }

  /** How far this file's newest revision has got, or null when it has none. */
  state(relPath: string): DraftState | null {
    const revision = this.latest.get(relPath);
    if (revision === undefined) return null;
    const acks = this.acksFor(relPath);
    const kept: Kept = acks.disk >= revision ? "disk" : acks.browser >= revision ? "browser" : "pending";
    return { revision, kept, trouble: this.troubles.get(relPath) ?? null };
  }

  /** The newest text alabs holds for this file, for handing to the user. */
  text(relPath: string): string | null {
    return this.live.get(relPath)?.contents ?? null;
  }

  /** Every file with unsaved work alabs is keeping right now. */
  paths(): string[] {
    return [...this.latest.keys()];
  }

  /**
   * That revision has been dealt with — written to the file, or discarded on
   * purpose — so the draft is no longer needed.
   *
   * A file edited again since then keeps its draft: both stores refuse to
   * forget text newer than the revision named here, so this can be called
   * after a save without checking first.
   */
  async settle(relPath: string, revision: number): Promise<void> {
    // Typed again since: that text has not been dealt with, so nothing is
    // forgotten and nothing is asked of either store.
    if ((this.latest.get(relPath) ?? 0) > revision) return;
    this.settled.set(relPath, Math.max(this.settled.get(relPath) ?? 0, revision));
    // Nothing still waiting is worth writing now; and a write already on its
    // way has to land before the record may be removed, or the drop could
    // overtake it and leave behind exactly the draft being settled.
    this.waiting.delete(relPath);
    await this.storing.get(relPath)?.catch(() => {});
    if ((this.latest.get(relPath) ?? 0) > revision) return;
    await Promise.allSettled([
      this.io.browser.drop(this.io.root, relPath, revision),
      this.io.disk.drop(this.io.root, relPath, revision),
    ]);
    // Typed while the stores were being asked: that text is unsaved again.
    if ((this.latest.get(relPath) ?? 0) > revision) return;
    // A store that could not forget it leaves a record of text that is now in
    // the file. That is offered for review at the next launch like any other
    // draft, which is safe; what would not be safe is alabs going on claiming
    // unsaved work for a buffer that has none.
    this.latest.delete(relPath);
    this.acked.delete(relPath);
    this.waiting.delete(relPath);
    this.live.delete(relPath);
    this.troubles.delete(relPath);
    this.changed(relPath);
  }
}

function sameStamp(a: FileStamp | null, b: FileStamp | null): boolean {
  if (a === null || b === null) return a === b;
  return a.identity === b.identity && a.mtime_secs === b.mtime_secs && a.mtime_nanos === b.mtime_nanos && a.len === b.len;
}

/**
 * The drafts to offer the user at launch: everything kept for this root,
 * newest first, taking the higher revision when the same file is in both
 * stores. Either can be the newer one — the browser is written first, and the
 * disk still holds what a cleared browser lost — so neither is trusted over
 * the other and the revision decides.
 */
export function mergeDrafts(browser: DraftListing, disk: DraftListing): DraftListing {
  const best = new Map<string, Draft>();
  for (const draft of [...disk.drafts, ...browser.drafts]) {
    const had = best.get(draft.relPath);
    if (!had || draft.revision > had.revision) best.set(draft.relPath, draft);
  }
  return {
    drafts: [...best.values()].sort((a, b) => b.updatedAt - a.updatedAt),
    unreadable: [...new Set([...browser.unreadable, ...disk.unreadable])].sort(),
  };
}

File: src-tauri/src/recovery.rs
//! Disk-backed recovery records for the browser runtime.
//!
//! The browser keeps an editor draft in IndexedDB the instant it is typed,
//! because that is the storage it can reach without leaving the page. That is
//! not the durable guarantee: clearing browser data, a new profile, or a
//! machine that never comes back to that profile would take it. So an
//! acknowledged draft is also an ordinary file here, under the browser
//! runtime's own state folder, where Finder and Terminal can reach it and
//! where clearing browser data cannot.
//!
//! What this module owns:
//!
//!   * one record per (root, path), named deterministically, carrying its own
//!     root and path so a name collision refuses instead of mixing two drafts;
//!   * restrictive permissions — 0700 on the folder, 0600 on every record —
//!     because a draft is the user's unsaved work, not shared state;
//!   * atomic replacement, so a record is either the previous revision or the
//!     new one and never a half-written file;
//!   * revisions that only ever move forward, so a late write cannot put an
//!     older draft back;
//!   * a size budget that **refuses** rather than evicting. Deleting somebody's
//!     unsaved work to make room for other unsaved work is not a recovery
//!     system. A refusal is visible and leaves the live text in the editor.
//!
//! Nothing here reads or writes inside the alabs root. Recovery is alabs'
//! own state, kept beside the disposable layout but never mixed with it: the
//! layout may be thrown away at any time, and these may not.

use std::fs;
use std::io::{self, Write};
use std::os::unix::fs::{OpenOptionsExt, PermissionsExt};
use std::path::{Path, PathBuf};
use std::sync::Mutex;
use std::time::{SystemTime, UNIX_EPOCH};

use serde::{Deserialize, Serialize};

use super::FileStamp;

/// The recovery folder inside the browser runtime's state folder. A folder of
/// its own, so "forget the layout" can never mean "forget the drafts".
const FOLDER: &str = "recovery";

/// Suffix of a record. Anything else in the folder is left alone.
const SUFFIX: &str = ".json";

/// Largest draft kept. The editor will not open a file past this, so a draft
/// of one cannot exceed it either.
const MAX_CONTENTS_BYTES: usize = 2 * 1024 * 1024;

/// Most the recovery folder may hold. Reaching it refuses the write, names
/// the reason, and removes nothing: the live text stays in the editor and the
/// user is told, rather than one draft being deleted to store another.
const MAX_TOTAL_BYTES: u64 = 64 * 1024 * 1024;

/// Format of the records this version writes. A record from a later version
/// is reported as unreadable rather than guessed at.
const FORMAT: u32 = 1;

// One server process owns recovery. Serialize the whole read/check/replace
// (and read/check/remove), including its temporary file, across HTTP workers.
// A request timeout cannot release this lock while its disk task still runs.
static RECORDS: Mutex<()> = Mutex::new(());

/// One editor draft as it is kept on disk.
///
/// Everything needed to reason about it later without the page that wrote it:
/// which root and path it belongs to, what the file looked like when the
/// buffer was last in step with the disk (`stamp`, absent when the file was
/// already gone), which revision of the draft this is, and which server
/// lifetime and editing session produced it.
#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Draft {
    /// Record format. Written as `FORMAT`; anything else is not read.
    pub(crate) version: u32,
    /// Canonical absolute path of the alabs root this draft belongs to.
    pub(crate) root: String,
    /// Path of the file inside that root.
    pub(crate) rel_path: String,
    /// The file's stamp when the buffer was last in step with the disk; null
    /// when the file did not exist then.
    pub(crate) stamp: Option<FileStamp>,
    /// Only ever increases for one (root, path).
    pub(crate) revision: u64,
    /// The `alabs-serve` lifetime that stored it.
    pub(crate) server_id: String,
    /// The editing session (one browser tab) that typed it.
    pub(crate) session_id: String,
    /// Milliseconds since the epoch, for showing the user how old a draft is.
    pub(crate) updated_at: u64,
    pub(crate) contents: String,
}

/// What one listing found: the drafts for the root that was asked about, and
/// the names of any records that could not be read. Unreadable records are
/// reported rather than skipped, because a recovery file alabs cannot parse
/// is exactly the thing the user needs to be told about.
#[derive(Serialize, Debug, Default)]
#[serde(rename_all = "camelCase")]
pub(crate) struct Listing {
    pub(crate) drafts: Vec<Draft>,
    pub(crate) unreadable: Vec<String>,
}

/// The recovery folder under one state folder.
pub(crate) fn dir_in(state_dir: &Path) -> PathBuf {
    state_dir.join(FOLDER)
}

/// 64-bit FNV-1a with a chosen offset basis. Two runs with different bases
/// give the 128 bits `record_name` needs.
fn fnv1a(bytes: &[u8], basis: u64) -> u64 {
    let mut hash = basis;
    for b in bytes {
        hash ^= *b as u64;
        hash = hash.wrapping_mul(0x0000_0100_0000_01b3);
    }
    hash
}

/// The record name for one (root, path).
///
/// Deterministic, so the same draft always lands on the same record and an
/// update replaces it rather than accumulating. Hashed rather than encoded
/// because a root path plus a file path easily exceeds what a filename may
/// hold. The record carries the root and path in full and every read checks
/// them, so the hash is an index and never the authority.
fn record_name(root: &str, rel_path: &str) -> String {
    let key = format!("{}\u{0}{}", root, rel_path);
    let bytes = key.as_bytes();
    format!(
        "{:016x}{:016x}{}",
        fnv1a(bytes, 0xcbf2_9ce4_8422_2325),
        fnv1a(bytes, 0x9ae1_6a3b_2f90_404f),
        SUFFIX
    )
}

/// Create the recovery folder if it is not there, reachable only by its owner.
///
/// The mode is set once, when alabs creates it, rather than on every write:
/// afterwards the folder is the user's, and a folder they have deliberately
/// made unwritable must stay that way so the refusal is real. Every record in
/// it is written 0600 regardless, which is what actually keeps a draft
/// private.
fn ensure_dir(dir: &Path) -> Result<(), String> {
    if dir.is_dir() {
        return Ok(());
    }
    fs::create_dir_all(dir).map_err(|e| format!("cannot create the recovery folder: {}", e))?;
    fs::set_permissions(dir, fs::Permissions::from_mode(0o700))
        .map_err(|e| format!("cannot secure the recovery folder: {}", e))
}

/// Read one record. `Ok(None)` means there is nothing at that name; an
/// unreadable or foreign-format record is an error naming why.
fn read_record(path: &Path) -> Result<Option<Draft>, String> {
    let text = match fs::read_to_string(path) {
        Ok(text) => text,
        Err(e) if e.kind() == io::ErrorKind::NotFound => return Ok(None),
        Err(e) => return Err(format!("cannot read a recovery file: {}", e)),
    };
    let draft: Draft =
        serde_json::from_str(&text).map_err(|_| "a recovery file is unreadable".to_string())?;
    if draft.version != FORMAT {
        return Err("a recovery file was written by a later version".to_string());
    }
    Ok(Some(draft))
}

/// Total size of the records in `dir`, ignoring anything that is not one.
fn folder_bytes(dir: &Path) -> u64 {
    let Ok(entries) = fs::read_dir(dir) else {
        return 0;
    };
    entries
        .flatten()
        .filter(|e| e.file_name().to_string_lossy().ends_with(SUFFIX))
        .filter_map(|e| e.metadata().ok())
        .filter(|m| m.is_file())
        .map(|m| m.len())
        .sum()
}

/// Milliseconds since the epoch, or 0 on a clock before it.
fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

/// Store one revision of one draft, replacing the record atomically.
///
/// Refused, without touching what is stored, when:
///
///   * the contents are larger than the editor would ever open;
///   * a record already there belongs to a different root or path (a hash
///     collision: two drafts are never merged into one record);
///   * a record already there holds a **newer** revision. An equal revision is
///     allowed, so a caller that is unsure whether its last write landed may
///     write it again;
///   * the folder is at its budget. Nothing is deleted to make room.
///
/// Returns the revision now stored.
pub(crate) fn put(dir: &Path, draft: &Draft) -> Result<u64, String> {
    let _records = RECORDS
        .lock()
        .map_err(|_| "recovery storage lock failed".to_string())?;
    if draft.contents.len() > MAX_CONTENTS_BYTES {
        return Err(format!(
            "this draft is larger than {} MB and cannot be kept",
            MAX_CONTENTS_BYTES / (1024 * 1024)
        ));
    }
    ensure_dir(dir)?;
    let name = record_name(&draft.root, &draft.rel_path);
    let path = dir.join(&name);
    let existing = read_record(&path)?;
    let existing_len = if let Some(old) = &existing {
        if old.root != draft.root || old.rel_path != draft.rel_path {
            return Err("another draft is already stored under that name".to_string());
        }
        if old.revision > draft.revision {
            return Err(format!(
                "a newer draft of {} is already kept",
                draft.rel_path
            ));
        }
        fs::metadata(&path).map(|m| m.len()).unwrap_or(0)
    } else {
        0
    };

    let text = serde_json::to_string(&Draft {
        version: FORMAT,
        updated_at: if draft.updated_at == 0 {
            now_ms()
        } else {
            draft.updated_at
        },
        ..draft.clone()
    })
    .map_err(|e| format!("cannot write a recovery file: {}", e))?;

    // The budget counts what the folder would hold afterwards: this record
    // replaces its own previous size rather than adding to it.
    let total = folder_bytes(dir);
    let after = total - existing_len.min(total) + text.len() as u64;
    if after > MAX_TOTAL_BYTES {
        return Err(format!(
            "alabs keeps at most {} MB of recovery drafts and is full. \
             Save or discard some drafts; nothing was deleted to make room.",
            MAX_TOTAL_BYTES / (1024 * 1024)
        ));
    }

    write_atomically(dir, &name, &text)?;
    Ok(draft.revision)
}

/// Write `text` as `name` in `dir`: a 0600 sibling, synced, then renamed over
/// the record. A crash leaves the previous revision or the new one.
fn write_atomically(dir: &Path, name: &str, text: &str) -> Result<(), String> {
    let tmp = dir.join(format!(".{}.{}.tmp", name, std::process::id()));
    let mut file = fs::OpenOptions::new()
        .write(true)
        .create(true)
        .truncate(true)
        .mode(0o600)
        .open(&tmp)
        .map_err(|e| format!("cannot write a recovery file: {}", e))?;
    let written = file
        .write_all(text.as_bytes())
        .and_then(|_| file.sync_all())
        .map_err(|e| format!("cannot write a recovery file: {}", e));
    drop(file);
    if let Err(e) = written {
        let _ = fs::remove_file(&tmp);
        return Err(e);
    }
    fs::rename(&tmp, dir.join(name)).map_err(|e| {
        let _ = fs::remove_file(&tmp);
        format!("cannot write a recovery file: {}", e)
    })
}

/// Every draft kept for `root`, newest first, plus the names of records that
/// could not be read. Drafts belonging to other roots are left untouched and
/// are not returned: one server owns one root.
pub(crate) fn list(dir: &Path, root: &str) -> Result<Listing, String> {
    let mut listing = Listing::default();
    let entries = match fs::read_dir(dir) {
        Ok(entries) => entries,
        Err(e) if e.kind() == io::ErrorKind::NotFound => return Ok(listing),
        Err(e) => return Err(format!("cannot read the recovery folder: {}", e)),
    };
    for entry in entries.flatten() {
        let name = entry.file_name().to_string_lossy().into_owned();
        if !name.ends_with(SUFFIX) || name.starts_with('.') {
            continue;
        }
        match read_record(&entry.path()) {
            Ok(Some(draft)) if draft.root == root => listing.drafts.push(draft),
            Ok(_) => {}
            Err(_) => listing.unreadable.push(name),
        }
    }
    listing
        .drafts
        .sort_by_key(|d| std::cmp::Reverse(d.updated_at));
    listing.unreadable.sort();
    Ok(listing)
}

/// Remove the record for one draft, but only when it is the one the caller
/// means and only up to the revision it has finished with.
///
/// A record holding a **newer** revision than `revision` is kept: the caller
/// confirmed an older save, and the newer text has not been dealt with.
/// Returns true when a record was removed.
pub(crate) fn drop_draft(
    dir: &Path,
    root: &str,
    rel_path: &str,
    revision: u64,
) -> Result<bool, String> {
    let _records = RECORDS
        .lock()
        .map_err(|_| "recovery storage lock failed".to_string())?;
    let path = dir.join(record_name(root, rel_path));
    let Some(existing) = read_record(&path)? else {
        return Ok(false);
    };
    if existing.root != root || existing.rel_path != rel_path {
        return Ok(false);
    }
    if existing.revision > revision {
        return Err(format!(
            "{} has been edited since that version; its draft was kept",
            rel_path
        ));
    }
    match fs::remove_file(&path) {
        Ok(()) => Ok(true),
        Err(e) if e.kind() == io::ErrorKind::NotFound => Ok(false),
        Err(e) => Err(format!("cannot remove a recovery file: {}", e)),
    }
}

#[cfg(test)]
#[path = "recovery_tests.rs"]
mod tests;
